### PR TITLE
Adding a Method to Check Naughtyness

### PR DESCRIPTION
**Pull request by Will-Hyde**
Adding an exposed sdk method that can check the naughtyness of a given piece of content.

method returns:
```
ContentModerationResponse {
  flagged: boolean;  // Whether or not the content has been flagged for explicit content or special characters
  reason: string;       // The reason the text has been flagged
  offendingTerms: string[];  //Any particularly offensive terms used
}
```

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -582,6 +586,36 @@ export interface UploadJobStatus {
   };
 }
 
+export interface ContentModerationResponse {
+  flagged: boolean;
+  reason: string;
+  offendingTerms: string[];
+}
+
+export interface ContentModeration {
```
Comment (JamesEarle) [inline src/types.ts:595]: Consider moving these to `src/api/types.ts` instead

File: src/api/actions/checkContentModeration.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+): Promise<ContentModerationResponse> => {
+  let moderation: Maybe<ContentModeration>;
+  let response: ContentModerationResponse = {
+    offendingTerms: [],
+    flagged: false,
+    reason: ""
+  }
+  try {
+    if (text.match(regex))
+    {
```
[the remark this answers is not in the document]
Comment (Will-Hyde) [reply src/api/actions/checkContentModeration.ts:22]: Sorry, must be some subconscious c# instincts escaping, I should know to put the curly bracket on the same line

File: src/types.ts [hunk 3]
```diff
@@ -582,6 +586,36 @@ export interface UploadJobStatus {
...
+  Classification: {
```
[the remark this answers is not in the document]
Comment (Will-Hyde) [reply src/types.ts:596]: The fields in the API response are PascalCase, and when I had this interface as camelcase, the fields would not map over.

File: src/api/actions/checkContentModeration.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+const reasons = {
+  specialCharacters: "Contains special characters.",
+  explicitContent: "Contains explicit content."
+}
+export const checkContentModeration = async (
+  apiUri: string,
+  text: string
+): Promise<ContentModerationResponse> => {
+  let moderation: Maybe<ContentModeration>;
+  let response: ContentModerationResponse = {
```
Comment (IsaiahDickison) [inline src/api/actions/checkContentModeration.ts:15]: I may be totally wrong but response could be const?

File: src/index.ts [hunk 5]
```diff
@@ -481,6 +482,10 @@ export const createInstance = (config: Config): Instance => {
         return apiClient.checkBulkUploadJob([jobId]);
       },
 
+      checkContentModeration: (text: string): Promise<ContentModerationResponse> => {
+        return apiClient.checkContentModeration(text);
```
[the remark this answers is not in the document]
Comment (Will-Hyde) [reply src/index.ts:486]: I dont think it needs it, none of the other exposed methods here are using it

File: src/api/actions/checkContentModeration.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
+import { Maybe } from "../../utilities";
+import { ContentModeration, ContentModerationResponse } from "../../types";
```
[the remark this answers is not in the document]
Comment (Will-Hyde) [reply src/api/actions/checkContentModeration.ts]: Moved

File: src/api/actions/checkContentModeration.ts (hunk 4 above)
Comment (Will-Hyde) [reply src/api/actions/checkContentModeration.ts]: Made object const

File: src/types.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Will-Hyde) [reply src/types.ts]: Created new types for ContentClassification and ContentCategoryScore

File: test/actions.test.ts [hunk 7]
```diff
@@ -205,4 +205,34 @@ describe("Test Custom SDK Logic", () => {
...
+      Just made a juug 
+      I got bands everywhere`
+      const sdkInstance = zNSSDK.createInstance(config);
+      const moderation = await sdkInstance.utility.checkContentModeration(sample);
+      expect(moderation.flagged).to.be.true;
+      expect(moderation.reason).to.equal('Contains explicit content.')
+    });
+
+    it("flags special characters", async () => {
+      var sample = `2 Chainz!`
```
[the remark this answers is not in the document]
Comment (Will-Hyde) [reply test/actions.test.ts:219]: Removed ability to accept spaces

**Review by Remscar — APPROVED** [review]
So we've run into the problem I always knew was going to happen; we have more than one API we're using.

I'm thinking we might want to split this logic into it's own folder, and not include it in the `api` folder. Instead we call it `content-moderation-api` or something (`utilities-api`?), and then rename `api` to `zns-api`.

Or we just leave it as is, but this will eventually lead to more code smell if we support more and more API's.

File: src/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Remscar) [reply src/index.ts:486]: Since `apiClient.checkContentModeration` returns a promise, you can just return that promise.

File: src/api/types.ts [hunk 8]
```diff
@@ -3,3 +3,33 @@ export interface UploadedFileDto {
   hash: string;
   url: string;
 }
+
+export interface ContentModerationResponse {
```
Comment (Remscar) [inline src/api/types.ts:7]: Since this is a type that the user may use, we should include it in the root `types.ts` file.

Comment (Will-Hyde) [reply src/api/types.ts:7]: Moving

**Pull request merged**